consensus/core/src/hashing/sighash.rs

```diff
@@ -0,0 +1,258 @@
+};
+
+use super::{sighash_type::SigHashType, HasherExtensions};
+
+#[derive(Default)]
+pub struct SigHashReusedValues {
+    previous_outputs_hash: Option<Hash>,
+    sequence_hash: Option<Hash>,
+    sig_op_counts_hash: Option<Hash>,
+    outputs_hash: Option<Hash>,
+}
+
+impl SigHashReusedValues {
+    pub fn new() -> Self {
+        Self { previous_outputs_hash: None, sequence_hash: None, sig_op_counts_hash: None, outputs_hash: None }
+    }
+}
+
+fn previous_output_hash(tx: &PopulatedTransaction, hash_type: SigHashType, reused_values: &mut SigHashReusedValues) -> Hash {
+    if hash_type.is_sighash_anyone_can_pay() {
+        return 0.into();
+    }
+
+    if let Some(previous_outputs_hash) = reused_values.previous_outputs_hash {
+        previous_outputs_hash
+    } else {
+        let mut hasher = TransactionSigningHash::new();
+        for input in tx.tx.inputs.iter() {
+            hasher.update(input.previous_outpoint.transaction_id.as_bytes());
+            hasher.write_u32(input.previous_outpoint.index);
+        }
+        let previous_outputs_hash = hasher.finalize();
+        reused_values.previous_outputs_hash = Some(previous_outputs_hash);
+        previous_outputs_hash
+    }
+}
+
+fn sequence_hash(tx: &PopulatedTransaction, hash_type: SigHashType, reused_values: &mut SigHashReusedValues) -> Hash {
+    if hash_type.is_sighash_single() || hash_type.is_sighash_anyone_can_pay() || hash_type.is_sighash_none() {
+        return 0.into();
+    }
+
+    if let Some(sequence_hash) = reused_values.sequence_hash {
+        sequence_hash
+    } else {
+        let mut hasher = TransactionSigningHash::new();
+        for input in tx.tx.inputs.iter() {
+            hasher.write_u64(input.sequence);
+        }
+        let sequence_hash = hasher.finalize();
+        reused_values.sequence_hash = Some(sequence_hash);
+        sequence_hash
+    }
+}
+
+fn sig_op_counts_hash(tx: &PopulatedTransaction, hash_type: SigHashType, reused_values: &mut SigHashReusedValues) -> Hash {
+    if hash_type.is_sighash_anyone_can_pay() {
+        return 0.into();
+    }
+
+    if let Some(sig_op_counts_hash) = reused_values.sig_op_counts_hash {
+        sig_op_counts_hash
+    } else {
+        let mut hasher = TransactionSigningHash::new();
+        for input in tx.tx.inputs.iter() {
+            hasher.write_u8(input.sig_op_count);
+        }
+        let sig_op_counts_hash = hasher.finalize();
+        reused_values.sig_op_counts_hash = Some(sig_op_counts_hash);
+        sig_op_counts_hash
+    }
+}
+
+fn payload_hash(tx: &PopulatedTransaction) -> Hash {
+    if tx.tx.subnetwork_id == SUBNETWORK_ID_NATIVE {
+        return 0.into();
+    }
+
+    // TODO: Right now this branch will never be executed, since payload is disabled
+    // for all non coinbase transactions. Once payload is enabled, the payload hash
+    // should be cached to make it cost O(1) instead of O(tx.inputs.len()).
```

Comment:
Since it only hashes the payload, technically the comment stating `O(tx.inputs.len())` is wrong